File: haskell-server/src/Bolour/Util/BlackWhite.hs
--
-- Copyright 2017-2020 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
--

{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE UndecidableInstances #-}
{-# LANGUAGE StandaloneDeriving #-}
{-# LANGUAGE DeriveFunctor #-}

module Bolour.Util.BlackWhite (
    BlackWhite(..)
  , isJustWhite
  , fromWhite
  , fromWhites
  , isWhite
  , isBlack
  , hasValue
  , toValueWithDefaults
  )
  where

import Data.Maybe (isNothing, isJust, fromJust)
import qualified Bolour.Util.Empty as Empty

-- | Black represents a value in an inactive or disabled location, for example,
--   the black square of a crossword puzzle; white represents a
--   value in a live location, which may be empty at the moment (represented
--   as Maybe).
data BlackWhite val = Black | White (Maybe val)
  deriving Functor

deriving instance (Eq val) => Eq (BlackWhite val)
deriving instance (Show val) => Show (BlackWhite val)

toValueWithDefaults :: BlackWhite val -> val -> val -> val
toValueWithDefaults bw blackDefault whiteDefault | isBlack bw = blackDefault
                           | Empty.isEmpty bw = whiteDefault
                           | otherwise = fromJust $ fromWhite bw

isJustWhite :: BlackWhite val -> Bool
isJustWhite Black = False
isJustWhite (White Nothing) = False
isJustWhite _ = True

fromWhite :: BlackWhite val -> Maybe val
fromWhite Black = Nothing
fromWhite (White Nothing) = Nothing
fromWhite (White x) = x

fromWhites :: [BlackWhite val] -> Int -> Int -> [Maybe val]
fromWhites line begin end =
  toMaybe <$> [begin .. end]
    where toMaybe i = fromWhite (line !! i)

instance Empty.Empty (BlackWhite val)
  where isEmpty bw =
          case bw of
          Black -> False
          White maybe -> isNothing maybe

isWhite :: BlackWhite val -> Bool
isWhite (White _) = True
isWhite Black = False

isBlack :: BlackWhite val -> Bool
isBlack (White _) = False
isBlack Black = True

hasValue :: BlackWhite val -> Bool
hasValue (White value) = isJust value
hasValue Black = False

File: scala-server/scalautil/src/main/scala/com/bolour/util/BlackWhite.scala
/*
 * Copyright 2017-2020 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/eagerwords.com/blob/master/LICENSE.md
 */

package com.bolour.util

/**
  * Abstraction for representing a value that may be disabled/inactive (black),
  * or enabled/active (white) but may be empty. Makes a clear distinction
  * between inactive and empty. Empty may be filled. Inactive may not.
  */
sealed abstract class BlackWhite[T] {
  def isWhite: Boolean
  def isBlack: Boolean
  def isEmpty: Boolean
  def hasValue: Boolean
  def fromWhite: Option[T]
  def map[R](f: T => R): BlackWhite[R]
  def toValueWithDefaults(blackDefault: T, whiteDefault: T): T
}

/**
  * An inactive/disabled object, e.g. a black square in a crossword puzzle.
  */
case class Black[T]() extends BlackWhite[T] {
  override def isWhite = false
  override def isBlack = true
  override def isEmpty = false
  override def hasValue = false
  override def fromWhite = None
  override def map[R](f: T => R) = Black[R]
  def toValueWithDefaults(blackDefault: T, whiteDefault: T) = blackDefault
}

/**
  * An enabled/active object that may nevertheless not have any value,
  * e.g., a white square in a crossword puzzle.
  */
case class White[T](value: Option[T]) extends BlackWhite[T] {
  override def isWhite = true
  override def isBlack = false
  override def isEmpty = value.isEmpty
  override def hasValue = value.isDefined
  override def fromWhite = value
  override def map[R](f: T => R) = White(value map f)
  def toValueWithDefaults(blackDefault: T, whiteDefault: T) = whiteDefault
}

object BlackWhite {
  def fromWhites[T](line: List[BlackWhite[T]], begin: Int, end: Int): List[Option[T]] =
    (begin to end).toList map { i => line(i).fromWhite }
}
